lib/DxilValidation/DxilContainerValidation.cpp

```diff
@@ -73,6 +78,522 @@ struct DiagRestore {
+  }
+  // Verify each resource table.
+  unsigned ResIndex = 0;
+  // CBV
+  VerifyResourceTable(DM.GetCBuffers(), ResIndex, PSVVersion);
+  // Sampler
+  VerifyResourceTable(DM.GetSamplers(), ResIndex, PSVVersion);
+  // SRV
+  VerifyResourceTable(DM.GetSRVs(), ResIndex, PSVVersion);
+  // UAV
+  VerifyResourceTable(DM.GetUAVs(), ResIndex, PSVVersion);
+}
+
+void PSVContentVerifier::VerifyEntryProperties(const ShaderModel *SM,
+                                               PSVRuntimeInfo0 *PSV0,
+                                               PSVRuntimeInfo1 *PSV1,
+                                               PSVRuntimeInfo2 *PSV2) {
+  PSVRuntimeInfo3 DMPSV;
+  memset(&DMPSV, 0, sizeof(PSVRuntimeInfo3));
+  hlsl::InitPSVRuntimeInfo(&DMPSV, &DMPSV, &DMPSV, &DMPSV, DM);
+  if (PSV1) {
+    // Init things not set in InitPSVRuntimeInfo.
+    DMPSV.ShaderStage = static_cast<uint8_t>(SM->GetKind());
+    DMPSV.SigInputElements = DM.GetInputSignature().GetElements().size();
+    DMPSV.SigOutputElements = DM.GetOutputSignature().GetElements().size();
+    DMPSV.SigPatchConstOrPrimElements =
+        DM.GetPatchConstOrPrimSignature().GetElements().size();
+    // Set up ViewID and signature dependency info
+    DMPSV.UsesViewID = DM.m_ShaderFlags.GetViewID() ? true : false;
+    DMPSV.SigInputVectors = DM.GetInputSignature().NumVectorsUsed(0);
+    for (unsigned streamIndex = 0; streamIndex < 4; streamIndex++)
+      DMPSV.SigOutputVectors[streamIndex] =
+          DM.GetOutputSignature().NumVectorsUsed(streamIndex);
+    if (SM->IsHS() || SM->IsDS() || SM->IsMS())
+      DMPSV.SigPatchConstOrPrimVectors =
+          DM.GetPatchConstOrPrimSignature().NumVectorsUsed(0);
+  }
+  bool Mismatched = false;
+  if (PSV2)
+    Mismatched = memcmp(PSV2, &DMPSV, sizeof(PSVRuntimeInfo2)) != 0;
+  else if (PSV1)
+    Mismatched = memcmp(PSV1, &DMPSV, sizeof(PSVRuntimeInfo1)) != 0;
+  else
+    Mismatched = memcmp(PSV0, &DMPSV, sizeof(PSVRuntimeInfo0)) != 0;
+
+  if (Mismatched) {
+    std::string Str;
+    raw_string_ostream OS(Str);
+    hlsl::PrintPSVRuntimeInfo(OS, &DMPSV, &DMPSV, &DMPSV, &DMPSV,
+                              static_cast<uint8_t>(SM->GetKind()),
+                              DM.GetEntryFunctionName().c_str(), "");
+    OS.flush();
+    std::string Str1;
+    raw_string_ostream OS1(Str1);
+    PSV.PrintPSVRuntimeInfo(OS1, static_cast<uint8_t>(PSVShaderKind::Library),
+                            "");
+    OS1.flush();
+    EmitMismatchError("PSVRuntimeInfo", Str, Str1);
+  }
+}
+
+void PSVContentVerifier::Verify() {
+  unsigned ValMajor, ValMinor;
+  DM.GetValidatorVersion(ValMajor, ValMinor);
+  unsigned PSVVersion = MAX_PSV_VERSION;
+  // Constraint PSVVersion based on validator version
+  if (DXIL::CompareVersions(ValMajor, ValMinor, 1, 1) < 0)
+    PSVVersion = 0;
+  else if (DXIL::CompareVersions(ValMajor, ValMinor, 1, 6) < 0)
+    PSVVersion = 1;
+  else if (DXIL::CompareVersions(ValMajor, ValMinor, 1, 8) < 0)
+    PSVVersion = 2;
```

Comment:
After this point, we should verify that the strides in the loaded PSV are the same as retrieved from `PSVInitInfo(PSVVersion).<record>Size()`.  That will require exposing those strides publicly from `DxilPipelineStateValidation`, since it does not do so now.

Reply:
Note that this will make it unnecessary to check all the cases below for presence of specific `PSVRuntimeInfoN` against `PSVVersion`, since those are just based on this stride.

Reply:
Updated.